zebra-chain/src/parallel/tree.rs

```diff
@@ -160,36 +191,106 @@ impl NoteCommitmentTrees {
     }
 
     /// Update the sapling note commitment tree.
+    #[allow(clippy::unwrap_in_result)]
     fn update_sapling_note_commitment_tree(
         mut sapling: Arc<sapling::tree::NoteCommitmentTree>,
         sapling_note_commitments: Vec<sapling::tree::NoteCommitmentUpdate>,
-    ) -> Result<Arc<sapling::tree::NoteCommitmentTree>, NoteCommitmentTreeError> {
+    ) -> Result<
+        (
+            Arc<sapling::tree::NoteCommitmentTree>,
+            Option<(sapling::tree::Node, bool, u16)>,
+        ),
+        NoteCommitmentTreeError,
+    > {
         let sapling_nct = Arc::make_mut(&mut sapling);
+        let distance_to_complete_subtree =
+            (1 << TRACKED_SUBTREE_HEIGHT) - sapling_nct.count() % (1 << TRACKED_SUBTREE_HEIGHT);
+        let should_find_subtree_roots =
+            sapling_note_commitments.len() as u64 >= distance_to_complete_subtree;
+
+        // MAX_BLOCK_BYTES is less than 2^21 bytes, notes are 2^5 bytes, and a full subtree has
+        // 2^16 notes. Blocks should not contain multiple level 16 roots.
+        let mut subtree_root = None;
 
-        for sapling_note_commitment in sapling_note_commitments {
-            sapling_nct.append(sapling_note_commitment)?;
+        if should_find_subtree_roots {
+            let starts_new_subtree =
+                sapling_note_commitments.len() as u64 != distance_to_complete_subtree;
+            let complete_subtree_note_in_block_idx = distance_to_complete_subtree as usize - 1;
```

Comment:
Will this subtraction ever be performed when `distance_to_complete_subtree` is zero?

This underflow could panic in dev builds, or produce incorrect results in release builds.

Reply:
(And the same for Orchard.)

Reply:
> Will this subtraction ever be performed when `distance_to_complete_subtree` is zero?

No, `distance_to_complete_subtree` or `2^16 - (position % 2^16)` where position is an integer will always be 1 or greater.

Reply:
That correctness constraint isn't obvious in the code, would you mind documenting it or refactoring it?
(I haven't reviewed the latest code yet, so it might already be clearer.)

Reply:
Refactored in https://github.com/ZcashFoundation/zebra/pull/7334/commits/0ae8a47fe8640eed8b132b656973120f259ad13d.